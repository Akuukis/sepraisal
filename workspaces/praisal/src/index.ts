export * from './Praisal'
export * from './PraisalManager'
export * from './models'
export * from './unzipCachedSbc'
