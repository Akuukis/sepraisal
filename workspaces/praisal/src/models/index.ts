
export * from './Cube.ts'
export * from './Blueprint.ts'
export * from './Component.ts'
export * from './Ingot.ts'
export * from './Ore.ts'
export * from './Orientation.ts'


export * from './Group.ts'
