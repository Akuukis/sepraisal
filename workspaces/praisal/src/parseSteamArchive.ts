import { Stream } from 'stream'
import * as unzip from 'unzip'


export const parseSteamArchive = async (readStream: Stream): Promise<string> => {
    let result: string

    return (new Promise<string>((resolve, reject) => {
        readStream
            .on('error', reject)
            .pipe(unzip.Parse())
                .on('entry', (entry: unzip.Entry) => {
                    try {
                        const fileName = entry.path
                        const chunks: Buffer[] = []

                        entry.on('data', (chunk: Buffer) => {
                            chunks.push(chunk)
                        })

                        entry.on('end', () => {
                            if(fileName.endsWith('.sbc')) {
                                result = Buffer.concat(chunks).toString()
                            }
                        })
                    } catch(err) {
                        reject(err)
                    }
                })
                .on('error', reject)
                .on('close', () => {
                    resolve(result)
                })
    }))

}
