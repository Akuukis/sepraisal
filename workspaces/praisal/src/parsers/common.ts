/**
 * Not the best config, but as close to xml2js as possible.
 */
export const PARSE_CONFIG = {
    attributeNamePrefix: '',
    attrNodeName: '$',
    ignoreAttributes: false,
    arrayMode: (tagName: string): boolean => tagName !== '$',
}
