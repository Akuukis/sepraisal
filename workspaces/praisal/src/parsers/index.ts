export * from './parseBlueprint'
export * from './parseBlueprintSbc'
export * from './parseComponentSbc'
export * from './parsePhysicalItemsSbc'
