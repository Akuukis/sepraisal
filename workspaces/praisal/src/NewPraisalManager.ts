import { BLOCK_GROUPS, VENDOR_MOD } from '@sepraisal/common'
import { readFileSync } from 'fs'
import { join } from 'path'

import { PraisalManager } from './PraisalManager'

export const VENDOR_DIR = join(__dirname, '..', 'vendor')

export const NewPraisalManager = (): () => Promise<PraisalManager> => {
    const cubeBlocksSbcs = [
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Armor.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Automation.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Communications.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Control.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Doors.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Energy.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Extras.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Gravity.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Interiors.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_LCDPanels.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Lights.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Logistics.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Mechanical.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Medical.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Production.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Symbols.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Thrusters.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Tools.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Utility.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Weapons.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Wheels.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Windows.sbc')],

        [VENDOR_MOD.DECORATIVE_1        , join(VENDOR_DIR, VENDOR_MOD.DECORATIVE_1      , 'CubeBlocks.sbc')],
        [VENDOR_MOD.DECORATIVE_2        , join(VENDOR_DIR, VENDOR_MOD.DECORATIVE_2      , 'CubeBlocks.sbc')],
        [VENDOR_MOD.ECONOMY             , join(VENDOR_DIR, VENDOR_MOD.ECONOMY           , 'CubeBlocks.sbc')],
        [VENDOR_MOD.FROSTBITE           , join(VENDOR_DIR, VENDOR_MOD.FROSTBITE         , 'CubeBlocks.sbc')],
        [VENDOR_MOD.SPARKSOFTHEFUTURE   , join(VENDOR_DIR, VENDOR_MOD.SPARKSOFTHEFUTURE , 'CubeBlocks.sbc')],
    ].map(([mod, path]) => [mod, readFileSync(path).toString()] as [VENDOR_MOD, string])
    const componentsSbc = readFileSync(join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'Components.sbc')).toString()
    const blueprintsSbc = readFileSync(join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'Blueprints.sbc')).toString()
    const physicalItemsSbc = readFileSync(join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'PhysicalItems.sbc')).toString()

    return async () => {
        const sepraisal = new PraisalManager()
        await sepraisal.addPhysicalItemsSbc(physicalItemsSbc, VENDOR_MOD.VANILLA)
        await sepraisal.addBlueprintsSbc(blueprintsSbc, VENDOR_MOD.VANILLA)
        await sepraisal.addComponentsSbc(componentsSbc, VENDOR_MOD.VANILLA)
        for(const [mod, cubeBlocksSbc] of cubeBlocksSbcs) await sepraisal.addCubeBlocksSbc(cubeBlocksSbc, mod)
        sepraisal.build()
        sepraisal.addGroups(BLOCK_GROUPS)

        return sepraisal
    }
}
