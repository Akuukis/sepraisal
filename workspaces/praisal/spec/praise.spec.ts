import { BLOCK_GROUPS } from '@sepraisal/common'
import { readFileSync } from 'fs'
import { join } from 'path'

import { PraisalManager } from '../src'


// tslint:disable: no-duplicate-string
const VENDOR_DIR = join(__dirname, '..', 'vendor')
const PREFABS_DIR = join(VENDOR_DIR, 'prefabs')
const MY_BLUEPRINTS_DIR = join(VENDOR_DIR, 'myBlueprints')


let sepraisal: PraisalManager
beforeEach(async () => {
    sepraisal = new PraisalManager()
    const cubeBlocksXml = readFileSync(join(VENDOR_DIR, 'CubeBlocks.sbc')).toString()
    const componentsXml = readFileSync(join(VENDOR_DIR, 'Components.sbc')).toString()
    const materialsXml = readFileSync(join(VENDOR_DIR, 'Blueprints.sbc')).toString()
    const physicalItemsXml = readFileSync(join(VENDOR_DIR, 'PhysicalItems.sbc')).toString()
    await sepraisal.addOres(physicalItemsXml)
    await sepraisal.addIngots(physicalItemsXml, materialsXml)
    await sepraisal.addComponents(materialsXml, componentsXml)
    await sepraisal.addCubes(cubeBlocksXml)
    sepraisal.addGroups(BLOCK_GROUPS)
})

describe('PraisalManager.praiseXml', () => {
    test('should succeed to praise AtmosphericLander', async () => {
        const praisal = await sepraisal.praiseXml(readFileSync(join(PREFABS_DIR, 'AtmosphericLander.sbc'), 'utf-8'))
    })
    test('should succeed to praise RespawnShip', async () => {
        const praisal = await sepraisal.praiseXml(readFileSync(join(PREFABS_DIR, 'RespawnShip.sbc'), 'utf-8'))
    })
    test('should succeed to praise dekartaTests', async () => {
        const praisal = await sepraisal.praiseXml(readFileSync(join(MY_BLUEPRINTS_DIR, 'dekartaTests.sbc'), 'utf-8'))
    })
    test('should succeed to praise EveryLargeBlockOnce', async () => {
        const praisal = await sepraisal.praiseXml(readFileSync(join(MY_BLUEPRINTS_DIR, 'EveryLargeBlockOnce.sbc'), 'utf-8'))
    })
})
