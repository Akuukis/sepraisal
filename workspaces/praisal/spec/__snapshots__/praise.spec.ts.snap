// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PraisalManager.praiseSbc should succeed to praise AtmosphericLander 1`] = `
Object {
  "DLCs": Array [],
  "DLCsCount": 0,
  "_revision": 0,
  "_version": 12,
  "blockCount": 261,
  "blockCountTotal": 261,
  "blockIntegrity": 1052364,
  "blockMass": 280699,
  "blockPCU": 1594,
  "blockTime": 2815,
  "blockVolume": 6578125,
  "blocks": Object {
    "<Vanilla Heavy Armor blocks>": 2,
    "<Vanilla Interior Light blocks>": 4,
    "<Vanilla Light Armor blocks>": 192,
    "<Vanilla Text Panel blocks>": 3,
    "<Vanilla Window blocks>": 8,
    "AirVent/": 1,
    "AirtightSlideDoor/LargeBlockSlideDoor": 1,
    "Assembler/LargeAssembler": 1,
    "BatteryBlock/LargeBlockBatteryBlock": 4,
    "Beacon/LargeBlockBeacon": 1,
    "CargoContainer/LargeBlockSmallContainer": 2,
    "Cockpit/CockpitOpen": 1,
    "Conveyor/LargeBlockConveyor": 3,
    "ConveyorConnector/ConveyorTube": 3,
    "CubeBlock/LargeInteriorPillar": 4,
    "CubeBlock/LargeSteelCatwalkPlate": 8,
    "Gyro/LargeBlockGyro": 1,
    "LandingGear/LargeBlockLandingGear": 4,
    "MedicalRoom/LargeMedicalRoom": 1,
    "MyProgrammableBlock/LargeProgrammableBlock": 1,
    "OreDetector/LargeOreDetector": 1,
    "OxygenGenerator/": 1,
    "OxygenTank/": 1,
    "Reactor/LargeBlockSmallGenerator": 1,
    "Refinery/LargeRefinery": 1,
    "Thrust/LargeBlockLargeAtmosphericThrust": 2,
    "Thrust/LargeBlockSmallAtmosphericThrust": 8,
    "TimerBlock/TimerBlockLarge": 1,
  },
  "componentCount": 16053,
  "componentMass": 280699,
  "componentTime": 25035,
  "componentVolume": 96705,
  "components": Object {
    "BulletproofGlass": 568,
    "Computer": 504,
    "Construction": 1581,
    "Detector": 20,
    "Display": 58,
    "Girder": 131,
    "InteriorPlate": 735,
    "LargeTube": 265,
    "Medical": 15,
    "MetalGrid": 362,
    "Motor": 3220,
    "PowerCell": 320,
    "RadioCommunication": 40,
    "Reactor": 100,
    "SmallTube": 372,
    "SteelPlate": 7762,
  },
  "flagsGreen": Array [],
  "flagsGreenCount": 0,
  "flagsRed": Array [],
  "flagsRedCount": 0,
  "flagsYellow": Array [],
  "flagsYellowCount": 0,
  "gridCount": 1,
  "gridSize": "Large",
  "gridStatic": false,
  "gridTitle": "AtmosphericLander",
  "height": 22.5,
  "ingotCount": 296582,
  "ingotMass": 296582,
  "ingotTime": 36131,
  "ingotVolume": 40611,
  "ingots": Object {
    "Cobalt": 1086,
    "Iron": 263525,
    "Nickel": 19900,
    "Silicon": 9271,
    "Silver": 800,
    "Stone": 2000,
  },
  "integrityPlanes": Object {
    "front": Any<Array>,
    "maxValue": 88735,
    "rotated": true,
    "side": Any<Array>,
    "top": Any<Array>,
  },
  "length": 37.5,
  "missingDefinitions": Object {
    "blocks": Object {},
    "components": Object {},
    "ingots": Object {},
    "ores": Object {},
  },
  "missingDefinitionsCount": 0,
  "oreCount": 453078,
  "oreMass": 453078,
  "oreVolume": 167639,
  "ores": Object {
    "Cobalt": 3620,
    "Iron": 376464,
    "Nickel": 49750,
    "Silicon": 13244,
    "Silver": 8000,
    "Stone": 2000,
  },
  "orientation": Orientation {
    "Forward": "Backward",
    "Up": "Up",
  },
  "sizeMB": 0.13,
  "thrustAtmospheric": Object {
    "Backward": 1296000,
    "Forward": 1296000,
    "Left": 1296000,
    "Right": 1296000,
    "Up": 12960000,
  },
  "thrustHydrogen": Object {},
  "thrustIon": Object {},
  "vanilla": true,
  "width": 27.5,
}
`;

exports[`PraisalManager.praiseSbc should succeed to praise AtmosphericLander 2`] = `
".....AA.AA.....
....AAAAAAA....
....AAAAAAA....
AAA.AAAAAAA.AAA
AAAAACAAABAAAAA
AAAAABABBBAAAAA
AAAAAAABAAAAAAA
AAAAAAAAAAAAAAA
...A..AAA..A...
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise AtmosphericLander 3`] = `
"..AAAA.....
.AAAAAAA...
.AAAAAAAA..
.AAAAAAAA..
.ABBBCAAA..
.ABBBBAA...
AAAABBAAA..
A..BAB.AAA.
A..AAA.A.AA
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise AtmosphericLander 4`] = `
"......AAA......
....AAAAAAA....
....ABAAABA....
AAAAABBABBAAAAA
AAAAAAAAAAAAAAA
AAAAACAAAAAAAAA
....AAAAAAA....
....AAAAAAA....
.....AAAAA.....
.......A.......
.......A.......
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise EveryLargeBlockOnce 1`] = `
Object {
  "DLCs": Array [],
  "DLCsCount": 0,
  "_revision": 0,
  "_version": 12,
  "blockCount": 122,
  "blockCountTotal": 122,
  "blockIntegrity": 1137248,
  "blockMass": 352951,
  "blockPCU": 5067,
  "blockTime": 2562,
  "blockVolume": 7375000,
  "blocks": Object {
    "<Vanilla Heavy Armor blocks>": 12,
    "<Vanilla Interior Light blocks>": 3,
    "<Vanilla Light Armor blocks>": 12,
    "<Vanilla Text Panel blocks>": 7,
    "<Vanilla Window blocks>": 2,
    "AirtightHangarDoor/": 1,
    "AirtightSlideDoor/LargeBlockSlideDoor": 1,
    "Assembler/LargeAssembler": 1,
    "BatteryBlock/LargeBlockBatteryBlock": 1,
    "Beacon/LargeBlockBeacon": 1,
    "ButtonPanel/ButtonPanelLarge": 1,
    "CameraBlock/LargeCameraBlock": 1,
    "CargoContainer/LargeBlockLargeContainer": 1,
    "CargoContainer/LargeBlockSmallContainer": 1,
    "Cockpit/LargeBlockCockpit": 1,
    "Cockpit/LargeBlockCockpitSeat": 1,
    "Cockpit/PassengerSeatLarge": 1,
    "Collector/Collector": 1,
    "Conveyor/LargeBlockConveyor": 1,
    "ConveyorConnector/ConveyorTube": 1,
    "ConveyorConnector/ConveyorTubeCurved": 1,
    "ConveyorSorter/LargeBlockConveyorSorter": 1,
    "CryoChamber/LargeBlockCryoChamber": 1,
    "CubeBlock/ArmorCenter": 1,
    "CubeBlock/ArmorCorner": 1,
    "CubeBlock/ArmorInvCorner": 1,
    "CubeBlock/ArmorSide": 1,
    "CubeBlock/LargeBlockInteriorWall": 1,
    "CubeBlock/LargeCoverWall": 1,
    "CubeBlock/LargeCoverWallHalf": 1,
    "CubeBlock/LargeInteriorPillar": 1,
    "CubeBlock/LargeRamp": 1,
    "CubeBlock/LargeStairs": 1,
    "CubeBlock/LargeSteelCatwalk": 1,
    "CubeBlock/LargeSteelCatwalk2Sides": 1,
    "CubeBlock/LargeSteelCatwalkCorner": 1,
    "CubeBlock/LargeSteelCatwalkPlate": 1,
    "Decoy/LargeDecoy": 1,
    "Door/": 1,
    "Drill/LargeBlockDrill": 1,
    "ExtendedPistonBase/LargePistonBase": 1,
    "Gyro/LargeBlockGyro": 1,
    "InteriorTurret/LargeInteriorTurret": 1,
    "LandingGear/LargeBlockLandingGear": 1,
    "LargeGatlingTurret/": 1,
    "LargeMissileTurret/": 1,
    "LaserAntenna/LargeBlockLaserAntenna": 1,
    "MedicalRoom/LargeMedicalRoom": 1,
    "MergeBlock/LargeShipMergeBlock": 1,
    "MotorAdvancedRotor/LargeAdvancedRotor": 1,
    "MotorAdvancedStator/LargeAdvancedStator": 1,
    "MotorRotor/LargeRotor": 1,
    "MotorStator/LargeStator": 1,
    "MyProgrammableBlock/LargeProgrammableBlock": 1,
    "OreDetector/LargeOreDetector": 1,
    "OxygenFarm/LargeBlockOxygenFarm": 1,
    "OxygenGenerator/": 1,
    "OxygenTank/": 1,
    "OxygenTank/LargeHydrogenTank": 1,
    "Passage/": 1,
    "PistonTop/LargePistonTop": 1,
    "Projector/LargeProjector": 1,
    "RadioAntenna/LargeBlockRadioAntenna": 1,
    "Reactor/LargeBlockLargeGenerator": 1,
    "Reactor/LargeBlockSmallGenerator": 1,
    "Refinery/Blast Furnace": 1,
    "Refinery/LargeRefinery": 1,
    "ReflectorLight/LargeBlockFrontLight": 1,
    "RemoteControl/LargeBlockRemoteControl": 1,
    "SensorBlock/LargeBlockSensor": 1,
    "ShipConnector/Connector": 1,
    "ShipGrinder/LargeShipGrinder": 1,
    "ShipWelder/LargeShipWelder": 1,
    "SmallMissileLauncher/LargeMissileLauncher": 1,
    "SolarPanel/LargeBlockSolarPanel": 1,
    "SoundBlock/LargeBlockSoundBlock": 1,
    "TerminalBlock/ControlPanel": 1,
    "Thrust/LargeBlockLargeAtmosphericThrust": 1,
    "Thrust/LargeBlockLargeHydrogenThrust": 1,
    "Thrust/LargeBlockLargeThrust": 1,
    "Thrust/LargeBlockSmallAtmosphericThrust": 1,
    "Thrust/LargeBlockSmallHydrogenThrust": 1,
    "Thrust/LargeBlockSmallThrust": 1,
    "TimerBlock/TimerBlockLarge": 1,
    "UpgradeModule/LargeEffectivenessModule": 1,
    "UpgradeModule/LargeEnergyModule": 1,
    "UpgradeModule/LargeProductivityModule": 1,
    "Warhead/LargeWarhead": 1,
    "Wheel/Wheel1x1": 1,
    "Wheel/Wheel3x3": 1,
    "Wheel/Wheel5x5": 1,
  },
  "componentCount": 20645,
  "componentMass": 352951,
  "componentTime": 45419,
  "componentVolume": 131691,
  "components": Object {
    "BulletproofGlass": 217,
    "Computer": 977,
    "Construction": 2557,
    "Detector": 56,
    "Display": 100,
    "Explosives": 6,
    "Girder": 36,
    "InteriorPlate": 1279,
    "LargeTube": 546,
    "Medical": 18,
    "MetalGrid": 969,
    "Motor": 1482,
    "PowerCell": 100,
    "RadioCommunication": 105,
    "Reactor": 2100,
    "SmallTube": 679,
    "SolarCell": 32,
    "SteelPlate": 8126,
    "Superconductor": 220,
    "Thrust": 1040,
  },
  "flagsGreen": Array [],
  "flagsGreenCount": 0,
  "flagsRed": Array [],
  "flagsRedCount": 0,
  "flagsYellow": Array [],
  "flagsYellowCount": 0,
  "gridCount": 4,
  "gridSize": "Large",
  "gridStatic": false,
  "gridTitle": "EveryLargeBlockOnce",
  "height": 40,
  "ingotCount": 411963,
  "ingotMass": 411963,
  "ingotTime": 85351,
  "ingotVolume": 62933,
  "ingots": Object {
    "Cobalt": 13307,
    "Gold": 1480,
    "Iron": 324887,
    "Magnesium": 12,
    "Nickel": 14651,
    "Platinum": 416,
    "Silicon": 4350,
    "Silver": 10860,
    "Stone": 42000,
  },
  "integrityPlanes": Object {
    "front": Any<Array>,
    "maxValue": 92658,
    "rotated": true,
    "side": Any<Array>,
    "top": Any<Array>,
  },
  "length": 55,
  "missingDefinitions": Object {
    "blocks": Object {},
    "components": Object {},
    "ingots": Object {},
    "ores": Object {},
  },
  "missingDefinitionsCount": 0,
  "oreCount": 934837,
  "oreMass": 934837,
  "oreVolume": 345890,
  "ores": Object {
    "Cobalt": 44357,
    "Gold": 148000,
    "Iron": 464124,
    "Magnesium": 1714,
    "Nickel": 36628,
    "Platinum": 83200,
    "Silicon": 6214,
    "Silver": 108600,
    "Stone": 42000,
  },
  "orientation": Orientation {
    "Forward": "Forward",
    "Up": "Up",
  },
  "sizeMB": 0.13,
  "thrustAtmospheric": Object {
    "Forward": 7128000,
  },
  "thrustHydrogen": Object {
    "Forward": 8280000,
  },
  "thrustIon": Object {
    "Forward": 4665600,
  },
  "vanilla": true,
  "width": 32.5,
}
`;

exports[`PraisalManager.praiseSbc should succeed to praise EveryLargeBlockOnce 2`] = `
"............AA........
............AAAAA.....
...........AABAAA.....
............AAAAA.....
............AAAA......
............AAAA......
............AAAA......
......AAA...AAA.......
..A...AAA...AA........
..A...AAA.AABAAA.AAAAA
AAAAAABBB.ABBAAAAAAAAA
AABAAABBBAAAABAAAAAAAA
.AAAAAAAAAAAA.AA.AAAAA
......AAAAAAAAAA.AAAAA
......AAAAAAAAA.......
......AAAAAAA.........
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise EveryLargeBlockOnce 3`] = `
"........A....
.......AAA...
.......AAAA..
.......AAAA..
.......AB....
.......ABA...
.......ABA...
......AABA...
......AAA....
...AAAAABAA..
AAAAAABBCBAAA
...AAABBCBAAA
.....AABBBAAA
......AAAAA..
......AAAA...
......AAAA...
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise EveryLargeBlockOnce 4`] = `
".........AA...........
.........AA...........
.........AA...........
......AAAAA..AAA......
......AAAAA..AAA......
......AAAAA..AAA......
......AACAAAAAAAAAA...
.....AAABBAAAAAAAAA...
..AAAAABBCBBAAAAAAAAAA
AAAAAAAAABBA.AAAAAABAA
AAAAA..AAA...AAAAAA...
.............AAA......
.............AAA......
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise RespawnShip 1`] = `
Object {
  "DLCs": Array [],
  "DLCsCount": 0,
  "_revision": 0,
  "_version": 12,
  "blockCount": 118,
  "blockCountTotal": 118,
  "blockIntegrity": 531409,
  "blockMass": 153949,
  "blockPCU": 1144,
  "blockTime": 1549,
  "blockVolume": 2343750,
  "blocks": Object {
    "<Vanilla Interior Light blocks>": 2,
    "<Vanilla Light Armor blocks>": 83,
    "<Vanilla Window blocks>": 2,
    "AirtightSlideDoor/LargeBlockSlideDoor": 1,
    "Assembler/LargeAssembler": 1,
    "CargoContainer/LargeBlockSmallContainer": 1,
    "Cockpit/CockpitOpen": 1,
    "Collector/Collector": 1,
    "Conveyor/LargeBlockConveyor": 1,
    "ConveyorConnector/ConveyorTube": 1,
    "ConveyorConnector/ConveyorTubeCurved": 1,
    "CubeBlock/LargeSteelCatwalk": 4,
    "GravityGenerator/": 1,
    "Gyro/LargeBlockGyro": 1,
    "MedicalRoom/LargeMedicalRoom": 1,
    "OxygenGenerator/": 1,
    "Reactor/LargeBlockSmallGenerator": 1,
    "Refinery/LargeRefinery": 1,
    "ShipConnector/Connector": 1,
    "Thrust/LargeBlockSmallThrust": 12,
  },
  "componentCount": 8236,
  "componentMass": 153949,
  "componentTime": 12432,
  "componentVolume": 41473,
  "components": Object {
    "BulletproofGlass": 40,
    "Computer": 399,
    "Construction": 1361,
    "Display": 30,
    "Girder": 10,
    "GravityGenerator": 6,
    "InteriorPlate": 472,
    "LargeTube": 139,
    "Medical": 15,
    "MetalGrid": 148,
    "Motor": 100,
    "Reactor": 100,
    "SmallTube": 198,
    "SteelPlate": 4258,
    "Thrust": 960,
  },
  "flagsGreen": Array [],
  "flagsGreenCount": 0,
  "flagsRed": Array [],
  "flagsRedCount": 0,
  "flagsYellow": Array [],
  "flagsYellowCount": 0,
  "gridCount": 1,
  "gridSize": "Large",
  "gridStatic": false,
  "gridTitle": "RespawnShip",
  "height": 12.5,
  "ingotCount": 164466,
  "ingotMass": 164466,
  "ingotTime": 45979,
  "ingotVolume": 21285,
  "ingots": Object {
    "Cobalt": 11364,
    "Gold": 1020,
    "Iron": 145748,
    "Nickel": 2290,
    "Platinum": 384,
    "Silicon": 830,
    "Silver": 830,
    "Stone": 2000,
  },
  "integrityPlanes": Object {
    "front": Any<Array>,
    "maxValue": 78561,
    "rotated": true,
    "side": Any<Array>,
    "top": Any<Array>,
  },
  "length": 12.5,
  "missingDefinitions": Object {
    "blocks": Object {},
    "components": Object {},
    "ingots": Object {},
    "ores": Object {},
  },
  "missingDefinitionsCount": 0,
  "oreCount": 442102,
  "oreMass": 442102,
  "oreVolume": 163578,
  "ores": Object {
    "Cobalt": 37880,
    "Gold": 102000,
    "Iron": 208211,
    "Nickel": 5725,
    "Platinum": 76800,
    "Silicon": 1186,
    "Silver": 8300,
    "Stone": 2000,
  },
  "orientation": Orientation {
    "Forward": "Left",
    "Up": "Up",
  },
  "sizeMB": 0.12,
  "thrustAtmospheric": Object {},
  "thrustHydrogen": Object {},
  "thrustIon": Object {
    "Backward": 345600,
    "Down": 691200,
    "Forward": 345600,
    "Left": 1382400,
    "Right": 691200,
    "Up": 691200,
  },
  "vanilla": true,
  "width": 12.5,
}
`;

exports[`PraisalManager.praiseSbc should succeed to praise RespawnShip 2`] = `
"AAAAA
ABBAA
BBBBB
ABBAA
..A..
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise RespawnShip 3`] = `
"....AAAAA..
AAAAAAAAAAA
AABAAAAABBC
AAAAAAAAAAA
......AA...
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise RespawnShip 4`] = `
".ABAA
.BBBA
ABBBA
.ABAA
.ABAA
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise dekartaTests 1`] = `
Object {
  "DLCs": Array [],
  "DLCsCount": 0,
  "_revision": 0,
  "_version": 12,
  "blockCount": 180,
  "blockCountTotal": 180,
  "blockIntegrity": 550400,
  "blockMass": 118000,
  "blockPCU": 180,
  "blockTime": 1632,
  "blockVolume": 4062500,
  "blocks": Object {
    "<Vanilla Light Armor blocks>": 164,
    "<Vanilla Window blocks>": 16,
  },
  "componentCount": 6740,
  "componentMass": 118000,
  "componentTime": 6740,
  "componentVolume": 31020,
  "components": Object {
    "BulletproofGlass": 2240,
    "Girder": 400,
    "SteelPlate": 4100,
  },
  "flagsGreen": Array [],
  "flagsGreenCount": 0,
  "flagsRed": Array [],
  "flagsRedCount": 0,
  "flagsYellow": Array [],
  "flagsYellowCount": 0,
  "gridCount": 1,
  "gridSize": "Large",
  "gridStatic": false,
  "gridTitle": "dekarta Tests",
  "height": 30,
  "ingotCount": 122100,
  "ingotMass": 122100,
  "ingotTime": 24585,
  "ingotVolume": 25654,
  "ingots": Object {
    "Iron": 88500,
    "Silicon": 33600,
  },
  "integrityPlanes": Object {
    "front": Any<Array>,
    "maxValue": 32500,
    "rotated": false,
    "side": Any<Array>,
    "top": Any<Array>,
  },
  "length": 32.5,
  "missingDefinitions": Object {
    "blocks": Object {},
    "components": Object {},
    "ingots": Object {},
    "ores": Object {},
  },
  "missingDefinitionsCount": 0,
  "oreCount": 174429,
  "oreMass": 174429,
  "oreVolume": 64539,
  "ores": Object {
    "Iron": 126429,
    "Silicon": 48000,
  },
  "orientation": Orientation {
    "Forward": "Forward",
    "Up": "Up",
  },
  "sizeMB": 0.05,
  "thrustAtmospheric": Object {},
  "thrustHydrogen": Object {},
  "thrustIon": Object {},
  "vanilla": true,
  "width": 10,
}
`;

exports[`PraisalManager.praiseSbc should succeed to praise dekartaTests 2`] = `
"BB..
BAAA
BAAA
BA..
BA..
BA..
BBAA
AAAA
AAA.
AAA.
AAA.
AA..
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise dekartaTests 3`] = `
"AAAAAAAAAAAAA
AAAAAAAAAAAAA
AAAAAAAAAAAAA
AAAAAAAAAAAAA
AAAAAAAAAAAAA
AAAAAAAAAAAAA
AAAAAAAAAAAAA
AAAAA....AAAA
.........AAAA
.........AAAA
.........AAAA
.........AAAA
"
`;

exports[`PraisalManager.praiseSbc should succeed to praise dekartaTests 4`] = `
"BBBBAAAAAAAAA
BAAAAAAAAAAAA
.AA....AAAAA.
.AA....AA....
"
`;
